diff --git a/pyproject.toml b/pyproject.toml
index 088cab32..0d486637 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,6 +1,20 @@
 [build-system]
-requires = ["poetry-core>=1.0.0", "poetry-dynamic-versioning"]
-build-backend = "poetry.core.masonry.api"
+requires = ["maturin>=0.14.8,<0.15", "poetry-dynamic-versioning"]
+build-backend = "maturin"
+
+[project]
+name = "eko"
+requires-python = ">=3.9"
+classifiers = [
+  "Programming Language :: Rust",
+  "Programming Language :: Python :: Implementation :: CPython",
+  "Programming Language :: Python :: Implementation :: PyPy",
+]
+version = "0.0.0"
+
+[tool.maturin]
+manifest-path = "crates/eko/Cargo.toml"
+python-packages = ["ekomark", "ekobox"]

 [tool.poetry]
 name = "eko"
@@ -129,6 +143,12 @@ asv-show = "asv show --config benchmarks/asv.conf.json"
 asv-clean = { "shell" = "rm -rf benchmarks/env benchmarks/html benchmarks/results" }
 asv = ["asv-run", "asv-publish", "asv-preview"]
 bump-version = "python crates/bump-versions.py"
+compile = "pip install -e crates/eko/"
+rdocs.cmd = "cargo doc --workspace --no-deps"
+rdocs.env = { RUSTDOCFLAGS = "--html-in-header crates/katex-header.html" }
+rdocs-view = "xdg-open target/doc/ekors/index.html"
+rdocs-clean = "rm -rf target/doc/"
+rtest = "cargo test --workspace"

 [tool.pytest.ini_options]
 testpaths = ['tests/', 'benchmarks/']
