diff --git a/src/eko/evolution_operator/__init__.py b/src/eko/evolution_operator/__init__.py
index f3c234f0..3980359e 100644
--- a/src/eko/evolution_operator/__init__.py
+++ b/src/eko/evolution_operator/__init__.py
@@ -3,7 +3,6 @@ r"""Contains the central operator classes.
 See :doc:`Operator overview </code/Operators>`.
 """

-import functools
 import logging
 import os
 import time
@@ -11,8 +10,9 @@ from dataclasses import dataclass
 from multiprocessing import Pool
 from typing import Dict, Tuple

+import ekors
 import numpy as np
-from scipy import integrate
+from scipy import LowLevelCallable, integrate

 from .. import basis_rotation as br
 from .. import scale_variations as sv
@@ -22,7 +22,7 @@ from ..io.types import EvolutionMethod, OperatorLabel
 from ..kernels import ev_method
 from ..matchings import Atlas, Segment
 from ..member import OpMember
-from .quad_ker import quad_ker_ad as quad_ker
+from .quad_ker import cb_quad_ker_qcd, cb_quad_ker_qed

 logger = logging.getLogger(__name__)

@@ -229,49 +229,6 @@ class Operator(sv.ScaleVariationModeMixin):
         """Return the evolution method."""
         return ev_method(EvolutionMethod(self.config["method"]))

-    def quad_ker(self, label, logx, areas):
-        """Return partially initialized integrand function.
-
-        Parameters
-        ----------
-        label: tuple
-            operator element pids
-        logx: float
-            Mellin inversion point
-        areas : tuple
-            basis function configuration
-
-        Returns
-        -------
-        functools.partial
-            partially initialized integration kernel
-        """
-        return functools.partial(
-            quad_ker,
-            order=self.order,
-            mode0=label[0],
-            mode1=label[1],
-            method=self.ev_method,
-            is_log=self.int_disp.log,
-            logx=logx,
-            areas=areas,
-            as_list=self.as_list,
-            mu2_from=self.q2_from,
-            mu2_to=self.q2_to,
-            a_half=self.a_half_list,
-            alphaem_running=self.alphaem_running,
-            nf=self.nf,
-            L=np.log(self.xif2),
-            ev_op_iterations=self.config["ev_op_iterations"],
-            ev_op_max_order=tuple(self.config["ev_op_max_order"]),
-            sv_mode=self.sv_mode,
-            is_threshold=self.is_threshold,
-            n3lo_ad_variation=self.config["n3lo_ad_variation"],
-            is_polarized=self.config["polarized"],
-            is_time_like=self.config["time_like"],
-            use_fhmruvv=self.config["use_fhmruvv"],
-        )
-
     def initialize_op_members(self):
         """Init all operators with the identity or zeros."""
         eye = OpMember(
@@ -294,6 +251,28 @@ class Operator(sv.ScaleVariationModeMixin):
             else:
                 self.op_members[n] = zero.copy()

+    def update_cfg(self, cfg):
+        """Adjust integration config."""
+        # prepare couplings for c
+        cfg.as1 = self.as_list[1]
+        cfg.as0 = self.as_list[0]
+        as_list_len = self.as_list.shape[0]
+        for j, a in enumerate(self.as_list.tolist()):
+            cfg.as_list[j] = a
+        cfg.as_list_len = as_list_len
+        a_half_x = self.a_half_list.shape[0]
+        a_half_y = self.a_half_list.shape[1]
+        for j, a in enumerate(self.a_half_list.flatten().tolist()):
+            cfg.a_half[j] = a
+        cfg.a_half_x = a_half_x
+        cfg.a_half_y = a_half_y
+
+        if self.order[1] == 0:
+            cfg.py = ekors.ffi.cast("void *", cb_quad_ker_qcd.address)
+        else:
+            cfg.py = ekors.ffi.cast("void *", cb_quad_ker_qed.address)
+        cfg.method_num = self.ev_method
+
     def run_op_integration(
         self,
         log_grid,
@@ -312,18 +291,62 @@ class Operator(sv.ScaleVariationModeMixin):
         """
         column = []
         k, logx = log_grid
+        # call(!) self.labels only once
+        labels = self.labels
         start_time = time.perf_counter()
+        # start preparing C arguments
+        cfg = ekors.lib.empty_args()
+        cfg.order_qcd = self.order[0]
+        cfg.order_qed = self.order[1]
+        cfg.is_polarized = self.config["polarized"]
+        cfg.is_time_like = self.config["time_like"]
+        cfg.nf = self.nf
+        cfg.is_log = self.int_disp.log
+        cfg.logx = logx
+        cfg.Lsv = np.log(self.xif2)
+        cfg.ev_op_iterations = self.config["ev_op_iterations"]
+        cfg.ev_op_max_order_qcd = self.config["ev_op_max_order"][0]
+        cfg.sv_mode_num = self.sv_mode
+        cfg.is_threshold = self.is_threshold
+        cfg.mu2_from = self.q2_from
+        cfg.mu2_to = self.q2_to
+        cfg.alphaem_running = self.alphaem_running
+        a_half_ffi = ekors.ffi.new("double[]", 2 * self.config["ev_op_iterations"])
+        cfg.a_half = a_half_ffi
+        as_list_ffi = ekors.ffi.new("double[]", self.config["ev_op_iterations"] + 1)
+        cfg.as_list = as_list_ffi
+        max_areas_shape = self.int_disp.max_areas_shape()
+        max_area_len = max_areas_shape[0] * max_areas_shape[1]
+        areas_ffi = ekors.ffi.new("double[]", max_area_len)
+        cfg.areas = areas_ffi
+
+        self.update_cfg(cfg)
+
         # iterate basis functions
         for j, bf in enumerate(self.int_disp):
             if k == j and j == self.grid_size - 1:
                 continue
+            # add emtpy labels with 0s
+            if bf.is_below_x(np.exp(logx)):
+                column.append({label: (0.0, 0.0) for label in labels})
+                continue
             temp_dict = {}
+            # prepare areas for C
+            curareas = bf.areas_representation
+            for j, x in enumerate(curareas.flatten().tolist()):
+                cfg.areas[j] = x
+            cfg.areas_x = curareas.shape[0]
+            cfg.areas_y = curareas.shape[1]
             # iterate sectors
-            for label in self.labels:
+            for label in labels:
+                cfg.mode0 = label[0]
+                cfg.mode1 = label[1]
+                # construct the low level object
+                func = LowLevelCallable(
+                    ekors.lib.rust_quad_ker, ekors.ffi.addressof(cfg)
+                )
                 res = integrate.quad(
-                    self.quad_ker(
-                        label=label, logx=logx, areas=bf.areas_representation
-                    ),
+                    func,
                     0.5,
                     1.0 - self._mellin_cut,
                     epsabs=1e-12,
