diff --git a/src/eko/evolution_operator/operator_matrix_element.py b/src/eko/evolution_operator/operator_matrix_element.py
index 9c41b1fa..c2e66e93 100644
--- a/src/eko/evolution_operator/operator_matrix_element.py
+++ b/src/eko/evolution_operator/operator_matrix_element.py
@@ -2,10 +2,10 @@
 evolution."""

 import copy
-import functools
 import logging
 from typing import Optional

+import ekors
 import numpy as np

 from .. import basis_rotation as br
@@ -13,8 +13,7 @@ from .. import scale_variations as sv
 from ..io.types import InversionMethod
 from ..matchings import Segment
 from . import Managers, Operator
-from .quad_ker import MatchingMethods
-from .quad_ker import quad_ker_ome as quad_ker
+from .quad_ker import MatchingMethods, cb_quad_ker_ome

 logger = logging.getLogger(__name__)

@@ -135,41 +134,15 @@ class OperatorMatrixElement(Operator):
             )
         return labels

-    def quad_ker(self, label, logx, areas):
-        """Return partially initialized integrand function.
-
-        Parameters
-        ----------
-        label: tuple
-            operator element pids
-        logx: float
-            Mellin inversion point
-        areas : tuple
-            basis function configuration
-
-        Returns
-        -------
-        functools.partial
-            partially initialized integration kernel
-        """
-        return functools.partial(
-            quad_ker,
-            order=self.order,
-            mode0=label[0],
-            mode1=label[1],
-            is_log=self.int_disp.log,
-            logx=logx,
-            areas=areas,
-            a_s=self.a_s,
-            nf=self.nf,
-            L=self.L,
-            sv_mode=self.sv_mode,
-            Lsv=np.log(self.xif2),
-            backward_method=self.backward_method,
-            is_msbar=self.is_msbar,
-            is_polarized=self.config["polarized"],
-            is_time_like=self.config["time_like"],
-        )
+    def update_cfg(self, cfg):
+        """Adjust integration config."""
+        cfg.is_ome = True
+        cfg.py = ekors.ffi.cast("void *", cb_quad_ker_ome.address)
+        cfg.L = self.L
+        cfg.as1 = self.a_s
+        cfg.as0 = 0.0
+        cfg.Lsv = np.log(self.xif2)
+        cfg.method_num = self.backward_method

     @property
     def a_s(self):
